src/core/experimental/array.d

```diff
@@ -0,0 +1,1549 @@
+     *  initialCapacity = The initial capacity for the array.
+     */
+    this(size_t initialCapacity)
+    {
+        rc = __RefCount.make!__RefCount();
+        reserve(initialCapacity);
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto a = rcarray!int(3);
+        assert(a.length == 0);
+        assert(a.capacity == 3);
+
+        a.insert([1, 2, 3]);
+        assert(a.length == 3);
+        assert(a.capacity == 3);
+    }
+
+
+    /**
+     * Creates an array out of the given _items.
+     *
+     * Params:
+     *  items = Any number of _items, in the form of a list of values, or a
+     *          built-in (static or dynamic) array.
+     */
+    this(U, this Q)(U[] items...)
+    if (!is(Q == shared) && is(U : T))
+    {
+        rc = __RefCount.make!__RefCount();
+
+        static if (is(Q == immutable) || is(Q == const))
+        {
+            mixin(immutableInsert!(typeof(items), "items")());
+        }
+        else
+        {
+            insert(items);
+        }
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        // Create a list from a list of ints
+        {
+            auto a = rcarray!int(1, 2, 3);
+            assert(a == [1, 2, 3]);
+        }
+        // Create a list from an array of ints
+        {
+            auto a = rcarray!int([1, 2, 3]);
+            assert(a == [1, 2, 3]);
+        }
+    }
+
+    private enum copyCtorIncRef = q{
+        // If we're not creating an immutable copy from non-immutable, we use
+        // the same underlying memory and increase the reference count (through
+        // the copy constructor of `__RefCount`)
+
+        rc = rhs.rc;
+        support = rhs.support;
+        payload = rhs.payload;
+    };
+
+    private enum copyCtorAlloc = q{
+        // If we are creating an immutable copy from non-immutable, we create a
+        // copy of the underlying memory and start a new reference count
+
+        rc = __RefCount.make!(immutable __RefCount)();
+
+        mixin(immutableInsert!(typeof(rhs.payload), "rhs.payload")());
+    };
+
+    this(return scope ref typeof(this) rhs)
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    this(return scope ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    this(return scope const ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    this(return scope immutable ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    this(return scope ref typeof(this) rhs) immutable
+    {
+        mixin(copyCtorAlloc);
+    }
+
+    this(return scope const ref typeof(this) rhs) immutable
+    {
+        mixin(copyCtorAlloc);
+    }
+
+    this(return scope immutable ref typeof(this) rhs) immutable
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    private this(RCQual, SuppQual, PaylQual, this Qualified)(RCQual _rc, SuppQual _support, PaylQual _payload)
+        if (is(typeof(support) : typeof(_support)))
```

Comment:
[DStyle](https://dlang.org/dstyle.html) is to put the contracts on the same indentation level.